examples/demos/retry_demo/main.go

```diff
@@ -0,0 +1,194 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"github.com/kentik/community_sdk_golang/apiv6/kentikapi/httputil"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"net/http/httptest"
+	"net/url"
+	"time"
+
+	"github.com/kentik/community_sdk_golang/examples/demos"
+	"github.com/kentik/community_sdk_golang/kentikapi"
+)
+
+func main() {
+	err := showRetryingOnMultipleCodes()
+	if err != nil {
+		log.Fatal(err)
+	}
+}
+
+func showRetryingOnMultipleCodes() error {
+	demos.Step("Create fake Kentik API server")
+	h := newSpyHTTPHandler([]httpResponse{
+		newErrorHTTPResponse(http.StatusBadGateway),
+		newErrorHTTPResponse(http.StatusBadGateway),
+		newErrorHTTPResponse(http.StatusBadGateway),
+		newErrorHTTPResponse(http.StatusBadGateway),
+		newErrorHTTPResponse(http.StatusServiceUnavailable),
+		newErrorHTTPResponse(http.StatusServiceUnavailable),
+		newErrorHTTPResponse(http.StatusTooManyRequests),
+		newErrorHTTPResponse(http.StatusTooManyRequests),
+		{
+			statusCode: http.StatusOK,
+			body:       dummyUsersResponseBody,
+		},
+	})
+	s := httptest.NewServer(h)
+	fmt.Printf("Running fake server on URL %v\n", s.URL)
+
+	demos.Step("Create Kentik API v5 client")
+	c := kentikapi.NewClient(kentikapi.Config{
+		APIURL: s.URL,
+		RetryCfg: httputil.RetryConfig{
+			MaxAttempts:          intPtr(42),
+			MinDelay:             durationPtr(100 * time.Millisecond),
+			MaxDelay:             durationPtr(10 * time.Second),
+			RetryableStatusCodes: []int{http.StatusTooManyRequests, http.StatusBadGateway, http.StatusServiceUnavailable},
+			RetryableMethods:     []string{"GET", "POST", "PUT", "PATCH", "DELETE"},
+		},
+	})
+
+	demos.Step("List users")
+	result, err := c.Users.GetAll(context.Background())
+
+	fmt.Println("Received result:")
+	demos.PrettyPrint(result)
+
+	return err
+}
+
+const (
+	retryAfterHeaderValue = "2"
+)
+
+type spyHTTPHandler struct {
+	// responses to return to the client
+	responses []httpResponse
+
+	// requests spied by the handler
+	requests []httpRequest
+}
+
+func newSpyHTTPHandler(responses []httpResponse) *spyHTTPHandler {
+	return &spyHTTPHandler{
+		responses: responses,
+	}
+}
+
+func (h *spyHTTPHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	body, err := ioutil.ReadAll(r.Body)
+	if err != nil {
+		writeResponse(
+			rw, http.StatusInternalServerError,
+			fmt.Sprintf("spyHTTPHandler: failed to read request body: %v", err),
+		)
+	}
+
+	err = r.Body.Close()
+	if err != nil {
+		log.Printf("spyHTTPHandler: failed to close request body: %v", err)
+	}
+
+	h.requests = append(h.requests, httpRequest{
+		method: r.Method,
+		url_:   r.URL,
+		header: r.Header,
+		body:   string(body),
+	})
+
+	rw.Header().Set("Content-Type", "application/json")
+	rw.Header().Set("Retry-After", retryAfterHeaderValue)
```

Comment:
We need to support (and test) also Retry-After with a timestamp in the values (i.e. the time to retry at) as opposed to just number of seconds.

This can be done later.

Reply:
> We need to support (and test) also Retry-After with a timestamp in the values (i.e. the time to retry at) as opposed to just number of seconds.
> 
> This can be done later.

We have separate ticket for that: https://github.com/kentik/community_sdk_golang/pull/50#discussion_r699700604